### PR TITLE
exit traffic via snapps

### PR DESCRIPTION
**Pull request by majestrate**
this makes the codepath for exit traffic via snapp work, adds new protocol type for exit traffic.

### CODE REVIEW

File: llarp/handlers/tun.cpp [hunk 1]
```diff
@@ -168,86 +164,12 @@ namespace llarp
...
        */
 
-#ifdef LOKINET_EXITS
-
-      if (not conf.m_exitNode.empty())
-      {
-        IPRange exitRange;
-        llarp::RouterID exitRouter;
-        std::string routerStr;
-        const auto pos = conf.m_exitNode.find(",");
```
Comment (notlesh) [inline llarp/handlers/tun.cpp:178]: I have a `split()` function in `util/str.hpp` which may not be quite as efficient but cuts down on a lot of verbose string parsing code (it uses string_view, but does allocate for `std::vector`)

File: CMakeLists.txt [hunk 2]
```diff
@@ -33,7 +33,7 @@ option(STATIC_LINK "link statically against dependencies" OFF)
 option(BUILD_SHARED_LIBS "build lokinet libraries as shared libraries instead of static" ON)
 option(SHADOW "use shadow testing framework. linux only" OFF)
 option(XSAN "use sanitiser, if your system has it (requires -DCMAKE_BUILD_TYPE=Debug)" OFF)
-option(JEMALLOC "use jemalloc. Not required on BSD" OFF)
+option(WITH_JEMALLOC "use jemalloc as allocator" OFF)
 option(TESTNET "testnet build" OFF)
```
Comment (notlesh) [inline CMakeLists.txt:37]: `jemalloc` could easily be its own commit or even PR

File: llarp/handlers/tun.cpp [hunk 3]
```diff
@@ -315,19 +237,9 @@ namespace llarp
     void
     TunEndpoint::Flush()
     {
-      static const auto func = [](auto self) {
-        self->FlushSend();
```
Comment (notlesh) [inline llarp/handlers/tun.cpp:319]: capture `this` ...?

File: CMakeLists.txt (hunk 2 above)
Comment (majestrate) [reply CMakeLists.txt:37]: yeah probably

File: llarp/net/ip_address.hpp [hunk 4]
```diff
@@ -21,6 +21,10 @@ namespace llarp
   {
     /// Empty constructor.
     IpAddress() = default;
+    /// move construtor
+    IpAddress(IpAddress&&) = default;
+    /// copy construct
+    IpAddress(const IpAddress&);
```
Comment (notlesh) [inline llarp/net/ip_address.hpp:27]: Can these other constructors be defaulted?

File: llarp/net/ip_address.hpp [hunk 5]
```diff
@@ -119,19 +131,7 @@ namespace llarp
       std::size_t
       operator()(const IpAddress& address) const noexcept
       {
-        (void)address;
-        // throw std::runtime_error("FIXME: IpAddress::Hash"); // can't do this in operator(),
-        // apparently, so hopefully it's me that stumbles upon this (if not, sorry!)
```
Comment (notlesh) [inline llarp/net/ip_address.hpp:124]: sorry! ;)

File: test/util/test_llarp_util_decaying_hashset.cpp [hunk 6]
```diff
@@ -2,11 +2,42 @@
...
+        hashset.Insert(rando, now);
+    }
+    if (now >= nextDecay)
+    {
+      REQUIRE(not hashset.Empty());
+      hashset.Decay(now);
+      nextDecay += decayInterval;
+    }
+
+  } while (llarp::time_now_ms() <= end);
```
Comment (notlesh) [inline test/util/test_llarp_util_decaying_hashset.cpp:33]: This seems like more of a stress test... a few of these aren't a big deal, but when you have a lot of tests that run for specific amounts of time, it can add up. Maybe we need a set of tests that only run on CI builds and/or as a special make target.

File: llarp/constants/path.hpp [hunk 7]
```diff
@@ -31,9 +31,9 @@ namespace llarp
     constexpr auto build_timeout = 30s;
 
     /// measure latency every this interval ms
-    constexpr auto latency_interval = 5s;
+    constexpr auto latency_interval = 20s;
```
Comment (notlesh) [inline llarp/constants/path.hpp:34]: This is a big change, would you mind briefly describing the rationale?

Comment (majestrate) [reply llarp/constants/path.hpp:34]: reduce idle bandwidth useage

File: llarp/util/str.cpp [hunk 8]
```diff
@@ -96,7 +96,7 @@ namespace llarp
         last = next;
 
         // advance to next non-delimiter
-        while (str[last] == delimiter)
+        while (str[last] == delimiter and last < str_size)
           last++;
```
Comment (notlesh) [inline llarp/util/str.cpp:100]: what were you running into here? i have a lot of tests around `split()` -- i'd be happy to add a regression test

File: llarp/router/i_outbound_message_handler.hpp [hunk 9]
```diff
@@ -24,9 +24,9 @@ namespace llarp
 
   using SendStatusHandler = std::function<void(SendStatus)>;
 
-  static const size_t MAX_PATH_QUEUE_SIZE = 40;
-  static const size_t MAX_OUTBOUND_QUEUE_SIZE = 200;
-  static const size_t MAX_OUTBOUND_MESSAGES_PER_TICK = 20;
+  static const size_t MAX_PATH_QUEUE_SIZE = 100;
+  static const size_t MAX_OUTBOUND_QUEUE_SIZE = 1000;
+  static const size_t MAX_OUTBOUND_MESSAGES_PER_TICK = 500;
```
Comment (notlesh) [inline llarp/router/i_outbound_message_handler.hpp:29]: as before -- what prompted these changes? just sheer bandwidth?

**Comment by majestrate** [comment]
On Thu, May 21, 2020 at 12:46:21PM -0700, Stephen Shelton wrote:
> @notlesh commented on this pull request.
> 
> 
> 
> > @@ -96,7 +96,7 @@ namespace llarp
>          last = next;
>  
>          // advance to next non-delimiter
> -        while (str[last] == delimiter)
> +        while (str[last] == delimiter and last < str_size)
>            last++;
> 
> what were you running into here? i have a lot of tests around `split()` -- i'd be happy to add a regression test 
out of bounds memory access potentially

**Comment by majestrate** [comment]
On Thu, May 21, 2020 at 12:47:33PM -0700, Stephen Shelton wrote:
> @notlesh commented on this pull request.
> 
> 
> 
> > @@ -24,9 +24,9 @@ namespace llarp
>  
>    using SendStatusHandler = std::function<void(SendStatus)>;
>  
> -  static const size_t MAX_PATH_QUEUE_SIZE = 40;
> -  static const size_t MAX_OUTBOUND_QUEUE_SIZE = 200;
> -  static const size_t MAX_OUTBOUND_MESSAGES_PER_TICK = 20;
> +  static const size_t MAX_PATH_QUEUE_SIZE = 100;
> +  static const size_t MAX_OUTBOUND_QUEUE_SIZE = 1000;
> +  static const size_t MAX_OUTBOUND_MESSAGES_PER_TICK = 500;
> 
> as before -- what prompted these changes? just sheer bandwidth?

sheer bandwidth, they can be reverted at this commit if needed.

File: daemon/main.cpp [hunk 10]
```diff
@@ -196,8 +222,7 @@ main(int argc, char* argv[])
 
     if (genconfigOnly)
     {
-      llarp::ensureConfig(
-          llarp::GetDefaultDataDir(), llarp::GetDefaultConfigPath(), overwrite, opts.isRelay);
+      llarp::ensureConfig(basedir, fname, overwrite, opts.isRelay);
```
Comment (tewinget) [inline daemon/main.cpp:225]: I can't help but feel like this is wholly unrelated to exit changes, and yet I'm okay with it being in here...

File: llarp/exit/session.cpp [hunk 11]
```diff
@@ -99,7 +99,7 @@ namespace llarp
     bool
     BaseSession::CheckPathDead(path::Path_ptr, llarp_time_t dlt)
     {
-      return dlt >= 10s;
+      return dlt >= path::alive_timeout;
```
Comment (tewinget) [inline llarp/exit/session.cpp:102]: <3

Comment (tewinget) [reply llarp/exit/session.cpp:102]: though I'd note the arg name "dlt" isn't very descriptive.

File: llarp/handlers/tun.cpp (hunk 3 above)
Comment (tewinget) [reply llarp/handlers/tun.cpp:319]: looks like it's capturing a shared_from_this using std::bind below

File: llarp/iwp/session.cpp [hunk 12]
```diff
@@ -615,15 +612,17 @@ namespace llarp
...
-          m_Parent->logic(), std::bind(&Session::HandlePlaintext, shared_from_this(), recvMsgs));
+      LogicCall(m_Parent->logic(), [self = shared_from_this(), msgs = recvMsgs] {
+        self->HandlePlaintext(std::move(msgs));
+      });
     }
 
     void
     Session::HandlePlaintext(CryptoQueue_ptr msgs)
     {
-      for (auto& result : *msgs)
```
Comment (tewinget) [inline llarp/iwp/session.cpp:625]: why?

File: llarp/path/transit_hop.cpp [hunk 13]
```diff
@@ -32,7 +32,7 @@ namespace llarp
       return stream;
     }
 
-    TransitHop::TransitHop() : m_UpstreamGather(128), m_DownstreamGather(128)
```
Comment (tewinget) [inline llarp/path/transit_hop.cpp:35]: magic numbers

File: llarp/service/endpoint.cpp [hunk 14]
```diff
@@ -636,8 +636,7 @@ namespace llarp
     std::set<RouterID>
     Endpoint::GetExitRouters() const
```
Comment (tewinget) [inline llarp/service/endpoint.cpp:637]: to be removed.

File: llarp/service/endpoint.hpp [hunk 15]
```diff
@@ -445,8 +448,10 @@ namespace llarp
       const ConvoMap& Sessions() const;
       ConvoMap&       Sessions();
       // clang-format on
-
+     protected:
```
Comment (tewinget) [inline llarp/service/endpoint.hpp:451]: maybe just move this declaration above with the other protected declarations?

File: llarp/service/outbound_context.cpp [hunk 16]
```diff
@@ -271,6 +271,17 @@ namespace llarp
       if (m_LookupFails > 16 || m_BuildFails > 10)
         return true;
 
+      constexpr auto InboundTrafficTimeout = 5s;
```
Comment (tewinget) [inline llarp/service/outbound_context.cpp:274]: maybe not the best place for such a const to be declared?

File: llarp/util/aligned.hpp [hunk 17]
```diff
@@ -28,14 +28,9 @@ namespace llarp
...
   template <size_t sz>
-#ifdef _WIN32
-  // We CANNOT align on a 128-bit boundary, malloc(3C) on win32
-  // only hands out 64-bit aligned pointers
-  struct alignas(uint64_t) AlignedBuffer
-#else
   struct alignas(std::max_align_t) AlignedBuffer
-#endif
   {
+    static_assert(alignof(std::max_align_t) <= 16, "insane alignment");
```
Comment (tewinget) [inline llarp/util/aligned.hpp:33]: perhaps a slightly more descriptive message would be good here.  Granted this is compile-time and will point to this source line (and the context around it), but still...

File: llarp/util/decaying_hashset.hpp [hunk 18]
```diff
@@ -56,13 +48,36 @@ namespace llarp
         return m_CacheInterval;
       }
 
+      bool
+      Empty() const
```
Comment (tewinget) [inline llarp/util/decaying_hashset.hpp:52]: maybe IsEmpty() ?

Shouldn't be too ambiguous where it's used, but there is a nonzero chance it could be confused with the verb "empty".

Comment (tewinget) [reply llarp/util/decaying_hashset.hpp:52]: (though the stl standard seems to be that empty() is a query, not a command, so I guess this is fine)

File: llarp/util/thread/logic.cpp [hunk 19]
```diff
@@ -6,92 +6,54 @@
 
 namespace llarp
 {
-  Logic::Logic(size_t sz) : m_Thread(llarp_init_threadpool(1, "llarp-logic", sz))
+  Logic::Logic(size_t)
```
Comment (tewinget) [inline llarp/util/thread/logic.cpp:9]: to be removed.

File: llarp/util/thread/logic.cpp [hunk 20]
```diff
@@ -6,92 +6,54 @@
...
-    LogDebug("logic thread spawned on ", spawned);
   }
 
   Logic::~Logic()
   {
-    delete m_Thread;
   }
 
   size_t
   Logic::numPendingJobs() const
```
Comment (tewinget) [inline llarp/util/thread/logic.cpp:18]: to be removed.

File: readme.md [hunk 21]
```diff
@@ -12,6 +12,8 @@ You can view documentation on how to get started [here](https://loki-project.git
 
 [![Build Status](https://drone.lokinet.dev/api/badges/loki-project/loki-network/status.svg?ref=refs/heads/master)](https://drone.lokinet.dev/loki-project/loki-network)
 
+You can find Bleeding edge builds [here](https://builds.lokinet.dev) (these builds may eat your first born).
```
Comment (tewinget) [inline readme.md:15]: Amusing, I suppose, but...probably not the *best* thing to have in the readme.

Comment (majestrate) [reply readme.md:15]: that was added in a commit to force ci to rebuild

File: llarp/path/transit_hop.cpp (hunk 13 above)
Comment (majestrate) [reply llarp/path/transit_hop.cpp:35]: fixed

File: llarp/util/thread/logic.cpp (hunk 19 above)
Comment (majestrate) [reply llarp/util/thread/logic.cpp]: fixed

File: llarp/util/thread/logic.cpp (hunk 20 above)
Comment (majestrate) [reply llarp/util/thread/logic.cpp]: fixed

File: readme.md (hunk 21 above)
Comment (majestrate) [reply readme.md]: removed

File: llarp/iwp/session.cpp (hunk 12 above)
Comment (majestrate) [reply llarp/iwp/session.cpp]: why indeed, reverted.

File: llarp/service/endpoint.cpp (hunk 14 above)
Comment (majestrate) [reply llarp/service/endpoint.cpp]: and it was. 
and tom saw it and saw that it was good.

File: llarp/context.cpp [hunk 22]
```diff
@@ -80,7 +77,12 @@ namespace llarp
     llarp::LogInfo(llarp::VERSION_FULL, " ", llarp::RELEASE_MOTTO);
     llarp::LogInfo("starting up");
     if (mainloop == nullptr)
-      mainloop = llarp_make_ev_loop();
+    {
+      auto jobQueueSize = config->router.m_JobQueueSize;
+      if (jobQueueSize < 1024)
```
Comment (tewinget) [inline llarp/context.cpp:82]: reference the new shiny constant here

File: llarp/ev/ev.cpp [hunk 23]
```diff
@@ -15,9 +15,9 @@
 #endif
 
 llarp_ev_loop_ptr
-llarp_make_ev_loop()
+llarp_make_ev_loop(size_t queueLength)
```
Comment (tewinget) [inline llarp/ev/ev.cpp:18]: pick a name and be consistent

File: llarp/ev/ev.h [hunk 24]
```diff
@@ -47,8 +49,9 @@ using llarp_ev_loop_ptr = std::shared_ptr<llarp_ev_loop>;
 
 /// make an event loop using our baked in event loop on Windows
 /// make an event loop using libuv otherwise.
+/// @param queue_size how big the logic job queue is
 llarp_ev_loop_ptr
-llarp_make_ev_loop();
+llarp_make_ev_loop(std::size_t queue_size = llarp::event_loop_queue_size);
```
Comment (tewinget) [inline llarp/ev/ev.h:54]: pick a name and be consistent

File: llarp/ev/ev_libuv.cpp [hunk 25]
```diff
@@ -753,7 +753,8 @@ namespace libuv
     llarp::LogContext::Instance().logStream->Tick(loop->time_now());
   }
 
-  Loop::Loop() : llarp_ev_loop(), m_LogicCalls(1024), m_timerQueue(20), m_timerCancelQueue(20)
+  Loop::Loop(size_t queueLength)
```
Comment (tewinget) [inline llarp/ev/ev_libuv.cpp:756]: pick a name and be consistent

File: llarp/ev/ev_libuv.hpp [hunk 26]
```diff
@@ -24,7 +24,7 @@ namespace libuv
       Callback callback;
     };
 
-    Loop();
+    Loop(size_t queueSize);
```
Comment (tewinget) [inline llarp/ev/ev_libuv.hpp:27]: pick a name and be consistent

File: llarp/service/outbound_context.cpp (hunk 16 above)
Comment (majestrate) [reply llarp/service/outbound_context.cpp:274]: i mean, it's used in only 1 place right now and making a header for all tunable parameters seems messier.

File: llarp/util/aligned.hpp [hunk 27]
```diff
@@ -28,14 +28,9 @@ namespace llarp
 {
   /// aligned buffer that is sz bytes long and aligns to the nearest Alignment
   template <size_t sz>
-#ifdef _WIN32
-  // We CANNOT align on a 128-bit boundary, malloc(3C) on win32
-  // only hands out 64-bit aligned pointers
-  struct alignas(uint64_t) AlignedBuffer
-#else
   struct alignas(std::max_align_t) AlignedBuffer
-#endif
```
Comment (despair86) [inline llarp/util/aligned.hpp:37]: why jeff why
